Update email requests

Messier-31 opened the pull request:
- Updated the email plugin settings to allow adding an alias
- Added a new email dialog 
- Added a nice looking html email with a QR code and payment info attached

Settings
![image](https://github.com/Electron-Cash/Electron-Cash/assets/82435373/6f0ca4ec-3b7d-4d66-98b1-e7c87c78be88)

Email Dialog
![image](https://github.com/Electron-Cash/Electron-Cash/assets/82435373/e48ab7c0-fff1-4adf-b92e-e8116a558b09)

Email size: 11.5Kb
![image](https://github.com/Electron-Cash/Electron-Cash/assets/82435373/84bc50a9-a2ad-4e78-af77-c6f8d86c2740)

cculianu commented:
Oh wow this is a cool update. Let me play with it and review it and get back to you. Thanks for the contribution!

cculianu commented on electroncash_gui/qt/request_list.py:
```diff
@@ -155,5 +157,5 @@ def create_menu(self, position):
         menu.addAction(_("Save as BIP70 file"), lambda: self.parent.export_payment_request(addr))
         menu.addAction(_("Delete"), lambda: self.parent.delete_payment_request(addr))
-        run_hook('receive_list_menu', menu, addr)
+        run_hook('receive_send_via_email_list_menu', menu, addr, self.parent.get_request_URI(addr), self.parent.receive_qr, status) 
```
Hmm. This hook seems oddly-specific. The hooks are intended to be generic mechanisms. 

I hesitate to change the old "hooks" API for this. Especially since the old hook was very generic and maybe some plugin out there would want to use it. I suggest we actually just update the API to more closely match what electrum is doing now -- they modified this hook to work as follows.

Also you don't really need to pass all that stuff in -- the callee can determine it anyway.

```suggestion
        run_hook('receive_list_menu', self.parent, menu, addr)
```

cculianu replied:
Ok, I suggest we keep the old hook in-place just for excessive caution. We then create a second hook tailored to this plugin. We also pass the "Window" (self.parent) in now, because it bothered me that the callee has to derive that.

```suggestion
        # Original old list menu API, here for legacy reasons
        run_hook('receive_list_menu', menu, addr)
        # New menu API specific to the email plugin
        run_hook('receive_list_menu_for_email_plugin', self.parent, menu, addr,
                 self.parent.get_request_URI(addr), self.parent.receive_qr, status)
```

cculianu commented on electroncash_plugins/email_requests/qt.py:
```diff
@@ -179,17 +392,21 @@ def new_invoice(self):
 
     @hook
-    def receive_list_menu(self, menu, addr):
+    def receive_send_via_email_list_menu(self, menu, addr, uri, qrcode, status):
         window = menu.parentWidget().parent  # Grr. Electrum programmers overwrote parent() method.
```
To match the API I proposed above:

```suggestion
    def receive_list_menu(self, window, menu, addr):
        uri = window.get_request_URI(addr)
        qrcode = window.receive_qr
```

cculianu replied:
And update the function name & signature:

```suggestion
    def receive_list_menu_for_email_plugin(self, window, menu, addr, uri, qrcode, status):
```

cculianu commented on electroncash_gui/qt/request_list.py:
```diff
@@ -131,6 +131,7 @@ def on_update(self):
             item = QTreeWidgetItem([date, address.to_ui_string(), '', message,
                                     amount_str, _(pr_tooltips.get(status,''))])
             item.setData(0, Qt.UserRole, address)
+            item.setData(1, Qt.UserRole, status)
```
I don't think this is needed. See my other comment.

```suggestion
```

cculianu commented on electroncash_gui/qt/request_list.py:
```diff
@@ -146,6 +147,7 @@ def create_menu(self, position):
             return
         self.setCurrentItem(item)  # sometimes it's not the current item.
         addr = item.data(0, Qt.UserRole)
+        status = item.data(1, Qt.UserRole)
```
I don't think this is needed...

```suggestion
```

cculianu commented on electroncash_plugins/email_requests/qt.py:
```diff
@@ -179,17 +392,21 @@ def new_invoice(self):
+    def receive_send_via_email_list_menu(self, menu, addr, uri, qrcode, status):
         window = menu.parentWidget().parent  # Grr. Electrum programmers overwrote parent() method.
-        menu.addAction(_("Send via e-mail"), lambda: self.send(window, addr))
+        menu.addAction(_("Send via e-mail"), lambda: self.open_email_dialog(window, addr, uri, qrcode, status))      
```
```suggestion
        menu.addAction(_("Send via e-mail"), lambda: self.open_email_dialog(window, addr, uri, qrcode))      
```

cculianu commented on electroncash_plugins/email_requests/qt.py:
```diff
@@ -179,17 +392,21 @@ def new_invoice(self):
+        menu.addAction(_("Send via e-mail"), lambda: self.open_email_dialog(window, addr, uri, qrcode, status))      
 
-    def send(self, window, addr):
+    def open_email_dialog(self, window, addr, uri ,qrcode, status):
```
```suggestion
    def open_email_dialog(self, window, addr, uri ,qrcode):
```

cculianu commented on electroncash_plugins/email_requests/qt.py:
```diff
@@ -179,17 +392,21 @@ def new_invoice(self):
         from electroncash import paymentrequest
         r = window.wallet.receive_requests.get(addr)
-        message = r.get('memo', '')
+        description = r.get('memo', '')
```
```suggestion
        status = r.get('status')
        description = r.get('memo', '')
```

cculianu reviewed (changes requested):
I am not sure about the proposed "hook" API update. See my suggestions (they are all related to that). 

- I don't think you need to pass all that info through the hook. It can be easily derived.
- I propose updating this "hook" API to math current Electrum (they pass: window, menu, addr)
- From that info, you can derive "status", "uri", and the "qrcode".

See my suggestions...

Messier-31 commented:
> Oh wow this is a cool update. Let me play with it and review it and get back to you. Thanks for the contribution!
@cculianu 

Hey, it's my pleasure to contribute and thank you for the feedback!

I apologize because I should have added some explanation of rationale for some of the decisions I had made.

I passed the qr-code and status through the hook for the following reasons;

1. The qr-code code is already generated so it's more efficient not to have to do it again, and;
2. The status can't be retrieved by the `receive_requests.get(addr)` object directly. In request_list.py the status is acquired by looping through the `self.wallet.get_sorted_requests(self.config)` object.  It's more efficient to attach the status to the tree item and then pass it though the hook rather than having to loop through the requests list again.

There are 3 options for status here;

1. Change the `receive_reqeusts.get()` functions to include "status" (though I can't seem to find where this function is), or;
2. Pass the status from request_list.py. (Like I have done), or;
3. Pass the "addr" over and loop through all the requests again. (Not efficient.)


I'm happy to make the changes you suggested, though I would ask that you consider the above rationale.

If in the case that you decide that it's ok to pass the qr-code and status through the hook, to address you concern of  genericity and interoperability, we can create an additional fallback hook in the name of the old one "receive_list_menu". Though I do  believe that this hook is only used in the right click receive list menu of the receive tab. Changing the name was only an attempt to make it easier to comprehend for someone new to the code.

Please advise accordingly

cculianu reviewed (changes requested):
Ok, so I request this change. Basically keep the old function in-place totally unchanged, just in case some *external* plugin should rely on it. New function is almost identical to what you proposed but with a changed name and also now takes "window" (ElectrumWindow instance) as its first passed-in argument.

cculianu commented:
You pushed a commit but it wasn't exactly what I suggested. Please fix?

Messier-31 commented:
> You pushed a commit but it wasn't exactly what I suggested. Please fix?
@cculianu 

I missed the comments. I'll do that now.

Also, before I do that - was the `self.parent` object you passed to both hooks a mistake?

cculianu commented:
> > You pushed a commit but it wasn't exactly what I suggested. Please fix?
> > @cculianu
> 
> I missed the comments. I'll do that now.
> 
> Also, before I do that - was the `self.parent` object you passed to both hooks a mistake?

1. It was only passed to the new hook.
2. It's needed because the hook is deriving it and that is less elegant than just passing it.

cculianu commented:
Hmm I don't have an imap server I can try this with sadly.  I tried to get gmail and yahoo and proton mail IMAP working but it fails. The little of the UI I do see looks good. This is likely an improvement. I will approve and merge this now because it looks good and is an improvement. ++ for using Qr Code.

If you have other enchancements like this you want to work on, I'm all eyes and ears.

Messier-31 commented:
> Hmm I don't have an imap server I can try this with sadly. I tried to get gmail and yahoo and proton mail IMAP working but it fails. The little of the UI I do see looks good. This is likely an improvement. I will approve and merge this now because it looks good and is an improvement. ++ for using Qr Code.
> 
> If you have other enchancements like this you want to work on, I'm all eyes and ears.

@cculianu 

I couldn't sleep thinking about some issues that the email plugin in has so I had to make an attempt to change it.

- All GUI windows would freeze while the thread was processing
- Changing any setting in the plugin demanded restarting the entire app
- Only one email dialog could be opened at once
- The previous user made some kind of polling request to the IMAP server trying to find any email that had a "bitcoin- paymentrequest " file, but nothing was done with them.
-I'm not entirely sure what the generated file is supposed to do (nothing happens when I open it with electron cash)

I've removed and changed a lot of the code now

- The SMTP request is now executing on a Qthread, Allowing use of other windows while dialog is open and/or sending.
- multiple dialogs can be opened and used at the same time.
- I removed all the IMAP code that was searching for request files.
- You don't have to restart the app to apply settings .
- Added some extra variables for the server in the settings.

Can you help me understand what the generated file is/does? I wasn't quite sure so I just left it in there.

I'll probably do some more work/testing on this before I make another pull request. Should I create a new branch for this or push it to this same pull request?

Unfortunately (or maybe fortunately) Google and most of the big email providers don't allow password logging anymore. We would have to set up Oauth as an alternative authentication method. I don't think it would be that hard. I would be up for it if the potential demand is there.

If you want to test it you can create an email with gmx.com. You'll just have to enable the imap/smtp settings in their webmail app.
